src/property.ts

```diff
@@ -0,0 +1,369 @@
+      maximum: this.maximum,
+      enum: this.enum,
+      readOnly: this.readOnly,
+      multipleOf: this.multipleOf,
+      links: this.links,
+      visible: this.visible,
+    };
+  }
+
+  /**
+   * @method isVisible
+   * @returns true if this is a visible property, which is a property
+   *          that is reported in the property description.
+   */
+  isVisible(): boolean {
+    return this.visible;
+  }
+
+  /**
+   * Make the property visible or invisible
+   */
+  setVisible(visible: boolean): void {
+    this.visible = visible;
+  }
+
+  isFireAndForget(): boolean {
+    return this.fireAndForget;
+  }
+
+  /**
+   * Sets the value and notifies the device if the value has changed.
+   * @returns true if the value has changed
+   */
+  setCachedValueAndNotify(value: T): boolean {
+    const oldValue = this.value;
+    this.setCachedValue(value);
+
+    // setCachedValue may change the value, therefore we have to check
+    // this.value after the call to setCachedValue
+    const hasChanged = oldValue !== this.value;
+
+    if (hasChanged) {
+      this.device.notifyPropertyChanged(this);
+    }
+
+    return hasChanged;
+  }
+
+  /**
+   * Sets this.value and makes adjustments to ensure that the value
+   * is consistent with the type.
+   */
+  setCachedValue(value: T): unknown {
+    if (this.type === 'boolean') {
+      // Make sure that the value is actually a boolean.
+      this.value = <T><unknown>!!value;
+    } else {
+      this.value = value;
+    }
+    return this.value;
+  }
+
+  /**
+   * @method getValue
+   * @returns a promise which resolves to the retrieved value.
+   *
+   * This implementation is a simple one that just returns
+   * the previously cached value.
+   */
+  getValue(): Promise<T> {
+    return new Promise((resolve) => {
+      if (this.value != this.prevGetValue) {
+        this.prevGetValue = this.value;
+      }
+      resolve(this.value);
+    });
+  }
+
+  /**
+   * @method setValue
+   * @returns a promise which resolves to the updated value.
+   *
+   * @note it is possible that the updated value doesn't match
+   * the value passed in.
+   *
+   * It is anticipated that this method will most likely be overridden
+   * by a derived class.
+   */
+  setValue(value: T): Promise<T> {
+    return new Promise((resolve, reject) => {
+      if (this.readOnly) {
+        reject('Read-only property');
+        return;
+      }
+
+      const numberValue = (<number> <unknown>value);
+
+      // eslint-disable-next-line no-undefined
+      if (this.minimum !== undefined && numberValue < this.minimum) {
+        reject(`Value less than minimum: ${this.minimum}`);
+        return;
+      }
+
+      // eslint-disable-next-line no-undefined
+      if (this.maximum !== undefined && numberValue > this.maximum) {
+        reject(`Value greater than maximum: ${this.maximum}`);
+        return;
+      }
+
+      // eslint-disable-next-line no-undefined
+      if (this.multipleOf !== undefined &&
```

Comment:
nit: I'd prefer all of the conditions to be aligned. Is that something we can fix with `eslint`? For example, I prefer this:

```js
if (condition1 &&
    condition2)
```

rather than this:

```js
if (condition1 &&
  condition2)
```

Reply:
I checked https://eslint.org/docs/rules/indent, but there seems to be no option for this.
The best thing I found was https://eslint.org/docs/rules/operator-linebreak, but this specifies only if the operator should be before or after the linebreak.